skiko/src/awtMain/objectiveC/macos/DisplayLinkThrottler.mm

```diff
@@ -21,125 +48,186 @@ static CVReturn displayLinkCallback(CVDisplayLinkRef displayLink, const CVTimeSt
+    [self updateDisplayLink];
     [self onVSync];
+
+    [_lock unlock];
 }
 
-- (void)systemDidWake:(NSNotification *)notification {
+- (void)systemDidWake {
+    [_lock lock];
+
     _isSleeping = NO;
+    [self updateDisplayLink];
+
+    [_lock unlock];
 }
 
 - (void)onVSync {
-    /// Lock condition lock and immediately unlock setting condition variable to 1 (can render now)
+    // Lock condition lock and immediately unlock setting condition variable to 1 (can render now)
     [_vsyncConditionLock lock];
     [_vsyncConditionLock unlockWithCondition:1];
 }
 
 - (void)waitVSync {
-    /// If display link is not constructed (due to failure or explicit opt-out), don't perform any waiting
-    if (!_displayLink) {
+    BOOL hasDisplayLink;
+
+    [_lock lock];
+    hasDisplayLink = _displayLink != nil;
+    [_lock unlock];
+
+    // No display link to wait for
+    if (!hasDisplayLink) {
         return;
     }
 
-    /// Wait until `onVSync` signals 1, then immediately lock, set it to 0 and unlock again.
+    // Wait until `onVSync` signals 1, then immediately lock, set it to 0 and unlock again.
     [_vsyncConditionLock lockWhenCondition:1];
     [_vsyncConditionLock unlockWithCondition:0];
 }
 
 - (void)invalidateDisplayLink {
+    // Assumed to be run inside _lock, except for dealloc
     if (_displayLink) {
         CVDisplayLinkStop(_displayLink);
         CVDisplayLinkRelease(_displayLink);
         _displayLink = nil;
-        _displayLinkScreen = nil;
-    }
-}
 
-- (void)setupDisplayLinkForWindow:(NSWindow *)window {
-    if (_isSleeping) {
-        /// System is sleeping, don't setup display link
-        return;
+        NSLog(@"DisplayLinkThrottler: Display link invalidated for screen %lld", _displayLinkScreenID);
     }
+}
 
-    NSScreen *screen = window.screen;
+- (void)createDisplayLink {
+    // Assumed to be run inside _lock
 
-    if (!screen) {
-        /// window is not yet attached to screen (or window is nil)
-        return;
-    }
+    [self invalidateDisplayLink];
 
-    if ([screen isEqualTo: _displayLinkScreen]) {
-        /// display link is already active for this screen, do nothing
+    if (_currentScreenID < 0) {
         return;
     }
 
-    [self invalidateDisplayLink];
-
-    NSDictionary* screenDescription = [screen deviceDescription];
-    NSNumber* screenID = [screenDescription objectForKey:@"NSScreenNumber"];
-
-    _displayLink = [self createDisplayLinkForScreen:screenID];
-    _displayLinkScreen = screen;
-}
-
-- (CVDisplayLinkRef)createDisplayLinkForScreen:(NSNumber *)screenID {
     CVReturn result;
-    CVDisplayLinkRef displayLink;
 
-    result = CVDisplayLinkCreateWithCGDisplay([screenID unsignedIntValue], &displayLink);
+    result = CVDisplayLinkCreateWithCGDisplay(
+        _currentScreenID /* truncated to uint32_t aka CGDirectDisplayID */,
+        &_displayLink
+    );
 
     if (result != kCVReturnSuccess) {
-        return nil;
+        _displayLink = nil;
```

Comment:
```suggestion
        _displayLink = nil;
        _displayLinkScreenID = nil;
```

Shouldn't we do this as well in all `return` points?

Reply:
No, and _displayLinkScreenID is uint32_t, we can't assign anything meaningful here.